src/mono/mono/utils/mono-mmap-wasm.c

```diff
@@ -96,125 +79,48 @@ mono_setmmapjit (int flag)
-	g_assert (!addr);
-	return mono_valloc_aligned (size, mono_pagesize (), flags, type);
-#else
-	return valloc_impl (addr, size, flags, type);
-#endif
+	g_assert (addr == NULL);
+	return mono_valloc_aligned (length, mono_pagesize (), flags, type);
 }
 
-static GHashTable *valloc_hash;
-
-typedef struct {
-	void *addr;
-	int size;
-} VallocInfo;
-
 void*
 mono_valloc_aligned (size_t size, size_t alignment, int flags, MonoMemAccountType type)
 {
-	// We don't need padding if the alignment is compatible with the page size
-	if (mono_opt_wasm_mmap && ((MWPM_PAGE_SIZE % alignment) == 0))
-		return valloc_impl (NULL, size, flags, type);
-
-	/* Allocate twice the memory to be able to put the block on an aligned address */
-	char *mem = (char *) valloc_impl (NULL, size + alignment, flags, type);
-	char *aligned;
+#ifdef DISABLE_THREADS
+	void *old_sbrk = NULL;
+	if ((flags & MONO_MMAP_NOZERO) == 0) {
+		old_sbrk = sbrk (0);
+	}
+#endif
 
-	if (!mem)
+	void *res = NULL;
+	if (posix_memalign (&res, alignment, size))
 		return NULL;
 
-	aligned = mono_aligned_address (mem, size, alignment);
-
-	/* The mmap implementation in emscripten cannot unmap parts of regions */
-	/* Free the other two parts in when 'aligned' is freed */
-	// FIXME: This doubles the memory usage
-	if (!valloc_hash)
-		valloc_hash = g_hash_table_new (NULL, NULL);
-	VallocInfo *info = g_new0 (VallocInfo, 1);
-	info->addr = mem;
-	info->size = size + alignment;
-	g_hash_table_insert (valloc_hash, aligned, info);
+#ifdef DISABLE_THREADS
+	if ((flags & MONO_MMAP_NOZERO) == 0 && old_sbrk > res) {
+		// this means that we got an old block, not the new block from sbrk
+		memset (res, 0, size);
+	}
+#else
+	if ((flags & MONO_MMAP_NOZERO) == 0) {
+		memset (res, 0, size);
+	}
+#endif
 
-	return aligned;
+	return res;
 }
 
 int
 mono_vfree (void *addr, size_t length, MonoMemAccountType type)
 {
-	VallocInfo *info = (VallocInfo*)(valloc_hash ? g_hash_table_lookup (valloc_hash, addr) : NULL);
-
-	if (info) {
-		/*
-		 * We are passed the aligned address in the middle of the mapping allocated by
-		 * mono_valloc_align (), free the original mapping.
-		 */
-		BEGIN_CRITICAL_SECTION;
-		if (mono_opt_wasm_mmap)
-			mwpm_free_range (info->addr, info->size);
-		else
-			munmap (info->addr, info->size);
-		END_CRITICAL_SECTION;
-		g_free (info);
-		g_hash_table_remove (valloc_hash, addr);
-	} else {
-		// FIXME: We could be trying to unmap part of an aligned mapping, in which case the
-		//  hash lookup failed because addr isn't exactly the start of the mapping.
-		// Ideally if the custom page manager is enabled, we won't have done aligned alloc.
-		BEGIN_CRITICAL_SECTION;
-		if (mono_opt_wasm_mmap)
-			mwpm_free_range (addr, length);
-		else
-			munmap (addr, length);
-		END_CRITICAL_SECTION;
-	}
-
-	mono_account_mem (type, -(ssize_t)length);
-
+	// NOTE: this doesn't implement partial freeing like munmap does
+	// we set MS_BLOCK_ALLOC_NUM to 1 to avoid partial freeing
+	g_free (addr);
```

Comment:
this is definitely an improvement over the old state of things, since g_free is dlfree and that performs way better than emscripten munmap